dvc/objects/db/__init__.py

```diff
@@ -77,4 +117,85 @@ def __getattr__(self, name):
 
     def by_scheme(self):
         self._init_odb(self.CLOUD_SCHEMES)
-        yield from self._odb.items()
+        for scheme in [Schemes.LOCAL] + self.CLOUD_SCHEMES:
+            yield scheme, self._odb[scheme]
+
+    def get_staging(
+        self, scheme: Optional[str] = None
+    ) -> Optional["ObjectDB"]:
+        return self._staging.get(scheme or Schemes.MEMORY)
+
+    def load_from_state(
```

Comment:
So one issue w/staging is that our local cache ODB and the local temp/staging ODB both share state.

In `objects.stage(odb, path_info, ...)` we used to check for a cached hash for `path_info` in `odb.state`, and then would try to load that object from `odb` (so we don't re-compute hashes for a path we've already staged).

This doesn't work when we are staging objects on one place and saving them elsewhere, we want `stage()` to load the object whether it's in the temp ODB or the main 'local cache' ODB, so staging has to be moved to a level that knows about state + all the possible ODBs where hashes from state can be stored.

It seemed to me that this fits at the ODBManager level, but I'm not sure if this should maybe go into a separate class for handling the "state + {set of ODBs that share state}" staging mechanic. (and then the repo ODBManager would also own an instance of that separate class)